compiler/rustc_ast/src/token.rs

```diff
@@ -905,6 +907,7 @@ impl NonterminalKind {
             },
             sym::pat_param => NonterminalKind::PatParam { inferred: false },
             sym::expr => NonterminalKind::Expr,
+            sym::expr_2021 if edition() >= Edition::Edition2024 => NonterminalKind::Expr2021,
```

Comment:
From [the RFC](https://github.com/rust-lang/rfcs/blob/master/text/3531-macro-fragment-policy.md#:~:text=In%20the%20current%20edition%2C%20the%20next%20edition%2C%20and%20as%20many%20other%20editions%20as%20practical%2C%20add%20a%20new%20fragment%20specifier%20that%20preserves%20the%20behavior%20of%20the%20existing%20fragment%20specifier.):

> In the current edition, the next edition, and as many other editions as practical, add a new fragment specifier that preserves the behavior of the existing fragment specifier.

My interpretation of this is that we need to allow `expr_2021` on Edition 2021, and if practical also support it on Edition 2018 and 2015. I don't see any reason why we shouldn't be able to support this on all editions.

Reply:
Actually, we probably can't do 2015 because `async` wasn't a keyword then.

Reply:
**TL;DR:** It's safe to assume that `expr_2021` will work for every previous edition until we encounter another `expr_xx`, such as `expr_2015` (if this were the case).

One practical idea to support this range of editions is that `expr_2021` remains backwards compatible until `expr_xx`. If there is another `expr_xx`, then the compatibility range includes everything.

However, consider if someone uses the `async` keyword in an edition like 2015; the code will likely complain that `async` is not supported, right? Perhaps not, if the keyword is used only inside the macros.

Reply:
This is fixed, as of [e0b7d92](https://github.com/rust-lang/rust/pull/123865/commits/e0b7d923b4900b382a2f20abb75a5aa000284542) we allow `expr_2021` on Edition 2021.